pkg/common/vsphere/cns_manager.go

```diff
@@ -0,0 +1,149 @@
+package vsphere
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/vmware/govmomi/cns"
+	cnstypes "github.com/vmware/govmomi/cns/types"
+	"github.com/vmware/govmomi/object"
+	"sync"
+)
+
+type CnsManager interface {
+	CreateVolume(ctx context.Context, createSpecList []cnstypes.CnsVolumeCreateSpec) (*object.Task, error)
+	DeleteVolume(ctx context.Context, volumeIDList []cnstypes.CnsVolumeId, deleteDisk bool) (*object.Task, error)
+	QueryVolume(ctx context.Context, queryFilter cnstypes.CnsQueryFilter) (*cnstypes.CnsQueryResult, error)
+	ResetManager(ctx context.Context, vcenter *VirtualCenter, cnsClient *cns.Client)
+}
+
+var (
+	cnsManagerInstance     *defaultCnsManager
+	cnsManagerInstanceLock sync.Mutex
+)
+
+// defaultCnsManager provides functionality to manage volumes.
+type defaultCnsManager struct {
+	logger        logrus.FieldLogger
+	virtualCenter *VirtualCenter
+	cnsClient     *cns.Client
+}
+
+// GetCnsManager returns the Manager instance.
+func GetCnsManager(ctx context.Context, vc *VirtualCenter, cnsClient *cns.Client, logger logrus.FieldLogger) (CnsManager, error) {
+	cnsManagerInstanceLock.Lock()
+	defer cnsManagerInstanceLock.Unlock()
+	if cnsManagerInstance != nil {
+		logger.Infof("Retrieving existing defaultCnsManager...")
+		cnsManagerInstance.virtualCenter = vc
+		cnsManagerInstance.cnsClient = cnsClient
+		logger.Infof("Updated existing defaultCnsManager...")
+		return cnsManagerInstance, nil
+	}
+	logger.Infof("Initializing new defaultCnsManager...")
+	cnsManagerInstance = &defaultCnsManager{
+		virtualCenter: vc,
+		cnsClient:     cnsClient,
+		logger:        logger,
+	}
+	return cnsManagerInstance, nil
+}
+
+func (m *defaultCnsManager) CreateVolume(ctx context.Context, createSpecList []cnstypes.CnsVolumeCreateSpec) (*object.Task, error) {
+	log := m.logger
+	invokeCount := 0
+	var createVolumeTask *object.Task
+	cachedCnsClient := m.cnsClient
+	var err error
+	for invokeCount <= DefaultAuthErrorRetryCount {
+		invokeCount++
+		createVolumeTask, err = cachedCnsClient.CreateVolume(ctx, createSpecList)
+		if err != nil {
```

Comment:
Can we put all of the error handling/reset connection stuff into a single function rather than having it duplicated everywhere?

Reply:
Done